src/nitypes/scalar.py

```
@@ -77,8 +91,15 @@ def __init__(
             raise invalid_arg_type("units", "str", units)
 
         self._value = value
-        self._extended_properties = ExtendedPropertyDictionary()
-        self._extended_properties[UNIT_DESCRIPTION] = units
+        if copy_extended_properties or not isinstance(
+            extended_properties, ExtendedPropertyDictionary
+        ):
+            extended_properties = ExtendedPropertyDictionary(extended_properties)
+        self._extended_properties = extended_properties
+        # If units are not already in extended properties, set them.
+        # If the caller specifies a non-blank units string, overwrite the existing entry.
+        if UNIT_DESCRIPTION not in self._extended_properties or units:
+            self._extended_properties[UNIT_DESCRIPTION] = units
```

It sounds like `Scalar(10, "Volts", extended_properties={"NI_UnitDescription": "Amps")` silently chooses Amps. I think it's better to return an error if units is truthy, extended_properties contains units, and they don't match.

---

I updated the logic for all 3 and added some units tests.